lib/vmfloaty/ssh.rb

```diff
@@ -14,27 +14,46 @@ def self.which(cmd)
     nil
   end
 
-  def self.command_string(verbose, service, host_os, use_token)
+  def self.command_string(verbose, service, host_os, use_token, ondemand = nil)
     ssh_path = which('ssh')
     raise 'Could not determine path to ssh' unless ssh_path
-
-    os_types = {}
+    os_types = Utils.generate_os_hash([host_os])
     os_types[host_os] = 1
 
-    response = service.retrieve(verbose, os_types, use_token)
+    response = service.retrieve(verbose, os_types, use_token, ondemand)
     raise "Could not get vm from #{service.type}:\n #{response}" unless response['ok']
 
     user = /win/.match?(host_os) ? 'Administrator' : 'root'
 
-    hostname = response[host_os]['hostname']
-    hostname = response[host_os]['hostname'][0] if response[host_os]['hostname'].is_a?(Array)
-    hostname = "#{hostname}.#{response['domain']}" unless hostname.end_with?('puppetlabs.net')
+    if ondemand
+      requestid = response['request_id']
+      service.wait_for_request(verbose, requestid)
+      response = service.retrieve(verbose, os_types, use_token, ondemand)
```

Comment:
I think this is superfluous since its already run on line 23.

Reply:
Ah yes thanks, forgot to remove during copy/paste